contracts/BandExampleConsumerToken.cdc

```diff
@@ -0,0 +1,194 @@
+        // elsewhere.
+        //
+        pub fun withdraw(amount: UFix64): @FungibleToken.Vault {
+            self.balance = self.balance - amount
+            emit TokensWithdrawn(amount: amount, from: self.owner?.address)
+            return <-create Vault(balance: amount)
+        }
+
+        // deposit
+        //
+        // Function that takes a Vault object as an argument and adds
+        // its balance to the balance of the owners Vault.
+        // It is allowed to destroy the sent Vault because the Vault
+        // was a temporary holder of the tokens. The Vault's balance has
+        // been consumed and therefore can be destroyed.
+        pub fun deposit(from: @FungibleToken.Vault) {
+            let vault <- from as! @BandExampleConsumerToken.Vault
+            self.balance = self.balance + vault.balance
+            emit TokensDeposited(amount: vault.balance, to: self.owner?.address)
+            vault.balance = 0.0
+            destroy vault
+        }
+
+        destroy() {
+            if self.balance > 0.0 {
+                BandExampleConsumerToken.totalSupply = BandExampleConsumerToken.totalSupply - self.balance
+            }
+        }
+    }
+
+    // createEmptyVault
+    //
+    // Function that creates a new Vault with a balance of zero
+    // and returns it to the calling context. A user must call this function
+    // and store the returned Vault in their storage in order to allow their
+    // account to be able to receive deposits of this token type.
+    //
+    pub fun createEmptyVault(): @FungibleToken.Vault {
+        return <-create Vault(balance: 0.0)
+    }
+
+    // Admin resource that allows to set the token price
+    pub resource Administrator {
+        
+        pub fun setNewTokensUSDPrice (price: UFix64) {
+            BandExampleConsumerToken.tokenUSDPrice = price
+            BandExampleConsumerToken.updateTokenFlowPrice()
+        }
+
+        pub fun updateTokenFlowPrice () {
+            BandExampleConsumerToken.updateTokenFlowPrice()
+        }
+
+    }
+
+    // Private contract function that calls the `BandOracle.getReferenceData` method
+    // for updating the `tokenFlowPrice` field.
+    access(contract) fun updateTokenFlowPrice() {
+        let payment <- self.flowTreasure.withdraw(amount: BandOracle.getFee())
+        let usdFlowData = BandOracle.getReferenceData (baseSymbol: "USD", quoteSymbol: "FLOW", payment: <- payment)
+        self.tokenFlowPrice = self.tokenUSDPrice * usdFlowData.fixedPointRate
+    }
+
+    // Public function that allows anyone to mint themselves a bunch of tokens, in 
+    // exchange of the needed amount of Flow tokens.
+    pub fun swapTokens(maxPrice: UFix64, payment: @FungibleToken.Vault): @BandExampleConsumerToken.Vault {
+        pre {
+            self.tokenFlowPrice < maxPrice: "Current token price is higher than the maximum desired price,"
+        }
+        let amount = payment.balance / self.tokenFlowPrice
```

Comment:
I think we'd need to denominate the token in USD and then perform this calculation. For instance, let's say `tokenFlowPrice` is `0.85`. If I create `MyToken.Vault` and mint myself `1_000_000.0` tokens, I could then mint myself `1_000_000.0 / 0.85` for nothing.

To avoid this, this method would need to get the quote for the token relative to Flow, so something like
`BandOracle.getReferenceData(baseSymbol: "FLOW", quoteSymbol: "<VAULT_SYMBOL>", payment: ...)`

This makes me think there should be a way to retrieve the quote symbol for a given type in the oracle contract. Something like

```
access(all) fun getQuoteSymbolByType(type: Type): String?
```

This would be especially helpful for consumer contracts that would like to get recent data on the fly in generalized cases like this one. Granted, this also means that the BandOracle admin would need to commit the vault type associated with a given symbol and checks added on updating data to ensure a symbol has an associated vault type. Not sure if that's feasible in the context of the offchain system.

Reply:
Thank you I totally skipped that!!! I'd say the most common sense fix here is just to cast the payment vault to a `BandExampleConsumerToken` vault, as you will do in any deposit method on a vault.
```
    // Public function that allows anyone to mint themselves a bunch of tokens, in 
    // exchange of the needed amount of Flow tokens.
    pub fun swapTokens(maxPrice: UFix64, payment: @FungibleToken.Vault): @BandExampleConsumerToken.Vault {
        pre {
            self.tokenFlowPrice < maxPrice: "Current token price is higher than the maximum desired price,"
        }
        let vault <- payment as! @BandExampleConsumerToken.Vault
        let amount = vault.balance / self.tokenFlowPrice
        self.flowTreasure.deposit(from: <-vault)
        self.totalSupply = self.totalSupply + amount
        emit TokensMinted(amount: amount)
        return <- create Vault(balance: amount)
    }
```

Regarding this swap method getting the quote straight from the oracle, we thought that won't be the best pattern since that will make you pay for the quotes every time a user calls the swap method, even if you are not getting new quotes.

Reply:
I find quite interesting your idea about allowing users to query the BandOracle using vault types, probably this should involve somehow the FT metadata views?